Implemented repeat attempts by default

danieldisu opened the pull request:
In this PR I've added a method that allows users to set repeat attempts by default, it works the same way than `allowFlakyAttemptsByDefault` but applies the `@Retry` annotation to all methods

Sloy reviewed (dismissed):
🙌

rocboronat commented:
Hi!

Wow, having a @Repeat is great! Just one thing. What do the tests do? I don't get the names: `allowFlakyStatementReturnedWhenAllowFlakyAnnotationFoundEvenRepeatStatement` and `lastStatementReturnedWhenRepeatSetRepeatAttemptsByDefaultAndAllowFlakyStatementAtTheSameTime`. In addition, the tests do and assert the same.

Thanks @danieldisu!

alorma reviewed (approved):
Good!!

danieldisu commented:
@rocboronat TBH Both tests were doing the same, I removed one and change the name of the other for `lastStatementReturnedWhenRepeatAttemptsByDefaultAndAllowFlakyStatementUsedAtTheSameTime`, I think this name it's more clear, but I'm open to suggestions!

> In addition, the tests do and assert the same.

I've extracted the Statement building to methods, but I'm not sure that's what you were pointing here 🤔

danieldisu commented:
@Sloy @rocboronat @alorma Can you review it again 🙏

rocboronat commented on library/src/main/java/com/schibsted/spain/barista/rule/flaky/FlakyTestRule.kt:
```diff
@@ -29,10 +29,23 @@ class FlakyTestRule : TestRule {
+   * Use this method when constructing the Rule to apply a default behavior of @[Repeat] without having to add the annotation to
+   * each test. This can help you to find flaky tests.
+   * <br></br>
+   * The default behavior can be overridden with [Repeat] or [Repeat].
```
`[Repeat] or [Repeat]`?

danieldisu replied:
It was `[Repeat] or [AllowFlaky]`, fixed!

rocboronat commented on library/src/test/java/com/schibsted/spain/barista/rule/flaky/FlakyStatementBuilderTest.java:
```diff
@@ -44,6 +44,16 @@ public void repeatStatementReturnedWhenRepeatAnnotationFound() throws Exception
   }
 
+  @Test
+  public void repeatStatementReturnedWhenRepeatSetRepeatAttemptsByDefault() throws Exception {
```
Sorry, I don't get the `repeat set repeat` 😅 

Maybe `repeatStatementReturnedWhenSettingDefaultRepeatAttempts`?

danieldisu replied:
Yep, it make sense 👍 , changed

rocboronat commented on library/src/test/java/com/schibsted/spain/barista/rule/flaky/FlakyStatementBuilderTest.java:
```diff
@@ -59,13 +69,36 @@ public void allowFlakyStatementReturnedWhenNoAnnotationsFoundButUsesDefault() th
+  }
+
+  @Test
+  public void lastStatementReturnedWhenRepeatAttemptsByDefaultAndAllowFlakyStatementUsedAtTheSameTime() throws Exception {
```
Not that important because it's easier to get, but just to be coherent:

`lastStatementReturnedWhenDefaultRepeatAttemptsAndAllowFlakyStatementUsedAtTheSameTime`

danieldisu replied:
Changed 👍

rocboronat commented:
There are some comments but if the Adevinta team wants to go on, it's good for me 👍 Thanks @danieldisu!

rocboronat reviewed (approved):
That's it then! Thanks!

rocboronat commented:
Let's wait for the CI to finish testing...
